### ISSUE TITLE
Simplifying matrices (arrays)

### ISSUE THREAD

**Issue by MarkusThill** [josdejong/mathjs#1913]
Hi there, 

I really love using this library, currently especially because of the simplify() function, since I can customize my own rules.
However, there is one point with the simplify() function that I did not manage to get working yet:

I have a set of rules which work fine for simple expressions, for example: "5x+3x" -> "8x".
Now I would like to apply these rules to every element of a matrix (ArrayNode). Unfortunately, this does not work. Obviously, I could myself extract each matrix element from the expression string and then simplify it. But it would be nicer if the simplify() function would do this for me. 
Is there a way to achieve this, which I missed? 

EDIT:
I actually managed to fix this, by adding a few lines of code to math.js myself. It was easier than I initially thought, since the code is quite clearly written. If there there is a less "invasive" way of solving the problem, I would be happy to hear about it. Otherwise this issue can be closed soon.

**Comment by josdejong** [comment]
Would be great to extend `simplify` with support for arrays. I suppose this means element-wise resolving of say constants? Like resolve `2 * [3x, 6x]` into `[6x, 12x]`? 

Can you share how you solved it now Markus?

Help implementing this would be very welcome.

**Comment by gwhitney** [comment]
Just noting that still (v10.0.2) `simplify` throws an error (`Unimplemented node type in simplifyConstant: ArrayNode`) when attempting to simplify any array (e.g. `math.simplify('[3x+5x]')`.

**Comment by gwhitney** [comment]
I am about to submit a pull request which addresses this issue. The one aspect mentioned here that it will not address is the suggestion that `2*[3x, 6x]` should simplify to `[6x, 12x]`. It is unclear to me which of these is simpler, or if `2*[3x, 6x]` should actually simplify to `6*[x, 2x]` or `6x*[1,2]`, say.

The upcoming PR will also not address the possibility that you might want `[x, 2x] + [3x, 4x]` to simplify to `[4x, 6x]`. It will handle `[1,2] + [3,4]` to `[4,6]` and `[x+3x, 2x+4x]` to `[4x, 6x]` and `x*[1,2]+x*[3,4]` to `[4,6]*x`, though. The difficulty is that the latter simplifications are handled by simplifyConstant, by simplifying the subexpressions in array entries, and by the existing rule to collect terms, respectively, but none of these apply to the former expression. And right now there is no notation for simplification rules that apply to matrices, so there's basically no way any other rules as they exist now could handle `[x, 2x] + [3x, 4x]`. One could imagine a version of the rules to collect like terms that would pull a scalar multiplier out of a matrix, but I don't see that there's any notation that capture that.

So once I file the PR, and presuming it is merged, you could either close this issue, keep it open for the sake of these other possible array simplifications, or you could close this and one of us could open a new issue for these possible more "advanced" array simplifications (although I think it is not always clear how you want to simplify array expressions with indeterminates...).

**Comment by gwhitney** [comment]
In particular, which is "simpler": `[a,b] + [c,d]` or `[a+c, b+d]`? Should we always merge sums of like-sized arrays? I doubt it, because I don't think `[1,2;-2,3] -  c*identity(2)` is simplified by changing it to `[1-c,2; -2, 3-c]`, is it? We might find some heuristics as to when merging sums of arrays or putting in or taking out a scalar factor are "simplifications" but I doubt there's a universal standard of "simplicity" here.  (Unlike, say, `15x^2` is definitely considered simpler than `3x*5x`.)

**Comment by josdejong** [comment]
Yes agree, I guess there is no single rule about what will be simplest in the end. 

This reminds me of units (see also https://github.com/ericman314/UnitMath), which have a similar issue: when is a simplification really a simplification. There it makes sense to calculate the "complexity" of the unit and see if it becomes simpler in the end when replacing some unit, or not. Would it be possible to somehow calculate the complexity of an expression too? Ie. `c*[a,b]` is less complex than `[c*a,c*b]` so better not write out `c` there, and `c*[a/c,b/c]` is more complex than `[a,b]`, so there it makes sense to write out `c`?

**Comment by josdejong** [comment]
Fixed in `v10.1.0` via #2382
